### PR TITLE
Closes #48469 - Refactor and DRY up Kahan Sum algorithm

### LINKED ISSUE #48469: Refactor and DRY up Kahan Sum algorithm
We use the Kahan summation algorithm in several places in aggregations, and while it's not that complicated of an algorithm, each repetition is still a chance to make a subtle error.  Most, if not all, of our usages could just call a common function (`GeoCentroidAggregator` might be the exception, since it runs a latitude and a longitude sum in parallel).  We should refactor the common case into one place and just call it as needed.

### PR DESCRIPTION
**Pull request by JosemyDuarte**
I have created `CompensatedSum` to centralize the logic of Kahan Sum and replace it in all the classes mentioned by #48469. 

I haven't put the looping logic inside the class since there are places where we do other stuff besides the sum in the loop, like in `StatsAggregator` where besides calculating the sum we also get the max and min value.

### CODE REVIEW

**Comment by not-napoleon** [comment]
@elasticmachine test this please

**Comment by JosemyDuarte** [comment]
Hi @not-napoleon! I just saw the build log that seems to be failing and I am not sure if I understand how my change could have caused that. If this is my fault I'm willing to solve it, but I would appreciate it if you can point me in the right direction.

Thanks!

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalWeightedAvg.java [hunk 1]
```diff
@@ -88,37 +88,21 @@ public String getWriteableName() {
...
-        double weightCompensation = 0;
+        CompensatedSum sumCompensation = CompensatedSum.newZeroInstance();
+        CompensatedSum weightCompensation = CompensatedSum.newZeroInstance();
+
         // Compute the sum of double values with Kahan summation algorithm which is more
         // accurate than naive summation.
         for (InternalAggregation aggregation : aggregations) {
             InternalWeightedAvg avg = (InternalWeightedAvg) aggregation;
-            // If the weight is Inf or NaN, just add it to the running tally to "convert" to
-            // Inf/NaN.  This keeps the behavior bwc from before kahan summing
```
Comment (not-napoleon) [inline server/src/main/java/org/elasticsearch/search/aggregations/metrics/InternalWeightedAvg.java]: I think this inline comment is worth preserving; Let's just copy it to the same clause in the new class.

**Review by not-napoleon — CHANGES_REQUESTED** [review]
I like the solution overall.  Left a couple of clean up comments that need to be fixed before we can merge it.

I appreciate you doing this work.  In the future though, please check if someone else has already asked to work on an issue before you start on it.

**Comment by not-napoleon** [comment]
I agree, the BWC failure looks unrelated.  I'll run it again after the requested changes land, and if it's green then we should be okay.

**Comment by JosemyDuarte** [comment]
Cool! Thanks for your feedback. I have a busy week, but I'll work on this before finishing the week.

**Comment by csoulios** [comment]
Just a quick note that Kahan summation is also used in #47468 for the `string_stats` aggregation

File: server/src/main/java/org/elasticsearch/search/aggregations/metrics/CompensatedSum.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+     * The correction term.
+     */
+    public double delta() {
+        return delta;
+    }
+
+    /**
+     * Increments the Kahan sum by adding a value and a correction term.
+     */
+    public CompensatedSum add(double value, double delta) {
```
[the remark this answers is not in the document]
Comment (JosemyDuarte) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/CompensatedSum.java]: Taking into consideration that we don't pass any delta different than 0, do you think it is better this one or the one which only takes `value` as a parameter (assuming delta as 0 always)?

[the remark this answers is not in the document]
Comment (JosemyDuarte) [reply server/src/main/java/org/elasticsearch/search/aggregations/metrics/CompensatedSum.java]: I kept both versions.

**Comment by not-napoleon** [comment]
@elasticmachine update branch

**Comment by not-napoleon** [comment]
@elasticmachine test this please

**Review by not-napoleon — APPROVED** [review]

**Comment by not-napoleon** [comment]
Hi @JosemyDuarte 

This looks good, thanks for making those changes.  I'm at a conference this week, so may be slow in checking in, but I'll keep an eye on the tests.  I'll let you know if there are failures I need you to look into.  Not sure I'll have time to merge it this week, but if I don't, I definitely will next week.

Thanks again for the contribution!

**Comment by JosemyDuarte** [comment]
Great! Thank you for your feedback. I'll keep an eye on simple issues to keep contributing. Enjoy your conference 👍

**Comment by not-napoleon** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-matrix

**Pull request merged**